### ISSUE TITLE
force-higher-level ML prediction: *** stack smashing detected ***

### ISSUE THREAD

**Issue by davidfrantz** [davidfrantz/force#176]
### Discussed in https://github.com/davidfrantz/force/discussions/173

<div type='discussions-op-text'>

<sup>Originally posted by **thielfab** March 21, 2022</sup>
I am running into the following error when running a ML prediction.
```
*** stack smashing detected ***: terminated
/var/spool/slurm/job623102/slurm_script: line 16: 1199735 Aborted                 (core dumped) /opt/singularity-3.8.2/bin/singularity exec ~/force_3.7.7.sif force-higher-level ~/project/higher_level/synthmix/ml.prm

```

Did you ever run into this issue? 

</div>

**Comment by davidfrantz** [comment]
Hi @thielfab,

it seems like only 8 bands are allocated for holding the prediction images. I see no obvious reason in the code why this would happen. Can you please share a sample dataset with me to reproduce and debug your issue on my end?

**Comment by davidfrantz** [comment]
Thanks Fabian,

I could reproduce your error and will debug this now.

It seems like the whole ``ml`` struct is gone in the 9th iteration. This is bad, and probably suggests that a rogue write is happening somewhere else in the code...

**Comment by davidfrantz** [comment]
I got it, and it was an embarassing mistake.

In the ML core routine, I needed to have two 2D arrays with ``number of modelsets`` x ``number of models within set``.
As the second one is variable, and I didn't want to bother with allocating a non-rectangular array, I apparently went for the easy but error-prone solution of using a static array of size 8 x 64...

I have now switched this to a dynamic array with ``number of modelsets`` x ``maximum number of models within the sets``.

https://github.com/davidfrantz/force/commit/71b41920eb84b24cf4b7ebb5a36249adb0e7f1f5

Cheers,
David
